node/core/pvf/src/executor_intf.rs

```diff
@@ -97,17 +100,61 @@ pub fn prevalidate(code: &[u8]) -> Result<RuntimeBlob, sc_executor_common::error
 
 /// Runs preparation on the given runtime blob. If successful, it returns a serialized compiled
 /// artifact which can then be used to pass into [`execute`] after writing it to the disk.
-pub fn prepare(blob: RuntimeBlob) -> Result<Vec<u8>, sc_executor_common::error::WasmError> {
-	sc_executor_wasmtime::prepare_runtime_artifact(blob, &CONFIG.semantics)
+pub fn prepare(
+	blob: RuntimeBlob,
+	ee_params: ExecutorParams,
+) -> Result<Vec<u8>, sc_executor_common::error::WasmError> {
+	let semantics = params_to_wasmtime_semantics(ee_params)
+		.map_err(|e| sc_executor_common::error::WasmError::Other(e))?;
+	sc_executor_wasmtime::prepare_runtime_artifact(blob, &semantics)
+}
+
+fn params_to_wasmtime_semantics(par: ExecutorParams) -> Result<Semantics, String> {
+	let mut sem = DEFAULT_CONFIG.semantics.clone();
+	let mut stack_limit = if let Some(stack_limit) = sem.deterministic_stack_limit.clone() {
+		stack_limit
+	} else {
+		return Err("No default stack limit set".to_string())
+	};
+	for p in par.iter() {
+		match p {
+			ExecutorParam::Environment(env) =>
+				if *env != ExecutionEnvironment::WasmtimeGeneric {
+					return Err("Wrong execution environment type".to_string())
+				},
+			ExecutorParam::ExtraHeapPages(ehp) => sem.extra_heap_pages = *ehp,
+			ExecutorParam::MaxMemorySize(mms) => sem.max_memory_size = Some(*mms as usize),
+			ExecutorParam::StackLogicalMax(slm) => stack_limit.logical_max = *slm,
+			ExecutorParam::StackNativeMax(snm) => stack_limit.native_stack_max = *snm,
+			ExecutorParam::InstantiationStrategy(is) => match is {
+				ExecInstantiationStrategy::PoolingCoW =>
+					sem.instantiation_strategy = InstantiationStrategy::PoolingCopyOnWrite,
+				ExecInstantiationStrategy::RecreateCoW =>
+					sem.instantiation_strategy = InstantiationStrategy::RecreateInstanceCopyOnWrite,
+				ExecInstantiationStrategy::Pooling =>
+					sem.instantiation_strategy = InstantiationStrategy::Pooling,
+				ExecInstantiationStrategy::Recreate =>
+					sem.instantiation_strategy = InstantiationStrategy::RecreateInstance,
+				ExecInstantiationStrategy::Legacy =>
+					sem.instantiation_strategy = InstantiationStrategy::LegacyInstanceReuse,
+			},
+			ExecutorParam::CanonicalizeNaNs(cnan) => sem.canonicalize_nans = *cnan,
+			ExecutorParam::ParallelCompilation(pc) => sem.parallel_compilation = *pc,
+			_ => (),
```

Comment:
nit: Not sure about Parity's stance on this yet, but I personally don't like having catch-all arms (`_`) when matching on an enum. If a new variant is ever added to the enum we won't get a compiler error that it's missing here.

Reply:
Neither do I love it in general, but in this very place it's on purpose. Imagine that we add another EE and we're writing completely different function that returns not Wasmtime `Semantics` built from `ExtraHeapPages` and `MaxMemorySize` etc, but some other structure build from a different set of `ExecutorParam` entries. The point is that it is safe for one EE to ignore parameters which are only relevant for another EE.

Reply:
Makes sense. Resolving.